includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php

```diff
@@ -789,69 +789,98 @@ public function process_link( $processed, $original ) {
-			];
-			if ( isset( $name_fragments[1] ) ) {
-				$merge_fields['LNAME'] = $name_fragments[1];
-			}
 			$update_payload = [
 				'email_address' => $email_address,
-				'merge_fields'  => $merge_fields,
 				'status'        => 'subscribed',
 			];
+			$merge_fields   = [];
+			if ( isset( $contact['name'] ) ) {
+				$name_fragments = explode( ' ', $contact['name'], 2 );
+				$merge_fields   = [
+					'FNAME' => $name_fragments[0],
+				];
+				if ( isset( $name_fragments[1] ) ) {
+					$merge_fields['LNAME'] = $name_fragments[1];
+				}
+				$update_payload['merge_fields'] = $merge_fields;
+			}
 
 			// Get list merge fields (metadata) to create them if needed.
-			$list_merge_fields = array_reduce(
-				$mc->get( "lists/$list_id/merge-fields" )['merge_fields'],
-				function( $acc, $field ) {
-					$acc[ $field['name'] ] = $field['tag'];
-					return $acc;
-				},
-				[]
-			);
-			foreach ( $contact['metadata'] as $key => $value ) {
-				if ( isset( $list_merge_fields[ $key ] ) ) {
-					$update_payload['merge_fields'][ $list_merge_fields[ $key ] ] = (string) $value;
-				} else {
-					$created_merge_field = $mc->post(
-						"lists/$list_id/merge-fields",
-						[
-							'name' => $key,
-							'type' => 'text',
-						]
-					);
-					$update_payload['merge_fields'][ $created_merge_field['tag'] ] = (string) $value;
+			if ( ! empty( $merge_fields ) ) {
+				$list_merge_fields = array_reduce(
+					$mc->get( "lists/$list_id/merge-fields" )['merge_fields'],
+					function( $acc, $field ) {
+						$acc[ $field['name'] ] = $field['tag'];
+						return $acc;
+					},
+					[]
+				);
+			}
+			if ( isset( $contact['metadata'] ) && is_array( $contact['metadata'] && ! empty( $contact['metadata'] ) ) ) {
+				foreach ( $contact['metadata'] as $key => $value ) {
+					if ( isset( $list_merge_fields[ $key ] ) ) {
+						$update_payload['merge_fields'][ $list_merge_fields[ $key ] ] = (string) $value;
+					} else {
+						$created_merge_field = $mc->post(
+							"lists/$list_id/merge-fields",
+							[
+								'name' => $key,
+								'type' => 'text',
+							]
+						);
+						$update_payload['merge_fields'][ $created_merge_field['tag'] ] = (string) $value;
+					}
 				}
 			}
 
 			// Create or update a list member.
 			$found_subscribers = $mc->get(
 				'search-members',
 				[
-					'list_id' => $list_id,
-					'query'   => $email_address,
+					'query' => $email_address,
 				]
 			)['exact_matches']['members'];
 			if ( empty( $found_subscribers ) ) {
-				$mc->post( "lists/$list_id/members", $update_payload );
+				$result = $mc->post( "lists/$list_id/members", $update_payload );
 			} else {
 				$member_id = $found_subscribers[0]['id'];
-				$mc->patch( "lists/$list_id/members/$member_id", $update_payload );
+				$result    = $mc->patch( "lists/$list_id/members/$member_id", $update_payload );
+			}
+			if (
+				! $result ||
+				( isset( $result['status'] ) && 400 === absint( $result['status'] ) ) ||
```

Comment:
How about

```suggestion
				( isset( $result['status'] ) && 'subscribed' === $result['status'] ) ||
```

since we only care about a successful subscription?

Also, `400` is [not the only possibility](https://mailchimp.com/developer/marketing/docs/errors/).

Reply:
Done in 308487e8444e0aa47411292b2205568e9946f1b0